agent/src/beerocks/slave/link_metrics/ieee802_3_link_metrics_collector.cpp

```diff
@@ -0,0 +1,246 @@
+
+    req.hdr.nlmsg_len    = NLMSG_LENGTH(sizeof(struct rtgenmsg));
+    req.hdr.nlmsg_type   = RTM_GETLINK;
+    req.hdr.nlmsg_flags  = NLM_F_REQUEST | NLM_F_DUMP;
+    req.hdr.nlmsg_seq    = 1;
+    req.hdr.nlmsg_pid    = 0;
+    req.gen.rtgen_family = AF_PACKET; /*  no preferred AF, we will get *all* interfaces */
+
+    io.iov_base          = &req;
+    io.iov_len           = req.hdr.nlmsg_len;
+    rtnl_msg.msg_iov     = &io;
+    rtnl_msg.msg_iovlen  = 1;
+    rtnl_msg.msg_name    = &kernel;
+    rtnl_msg.msg_namelen = sizeof(kernel);
+
+    if (sendmsg(fd, (struct msghdr *)&rtnl_msg, 0) < 0) {
+        LOG(ERROR) << "Unable to send message through Netlink socket: " << strerror(errno);
+    } else {
+        int msg_done = 0; /* flag to end loop parsing */
+
+        /**
+         * Parse reply until message is done
+         */
+        while (!msg_done) {
+            int length;
+            struct nlmsghdr *msg_ptr; /* pointer to current message part */
+
+            struct msghdr rtnl_reply; /* generic msghdr structure for use with recvmsg */
+            struct iovec io_reply;
+
+            /* a large buffer to receive lots of link information */
+            char reply[IFLIST_REPLY_BUFFER];
+
+            memset(&io_reply, 0, sizeof(io_reply));
+            memset(&rtnl_reply, 0, sizeof(rtnl_reply));
+
+            io.iov_base            = reply;
+            io.iov_len             = IFLIST_REPLY_BUFFER;
+            rtnl_reply.msg_iov     = &io;
+            rtnl_reply.msg_iovlen  = 1;
+            rtnl_reply.msg_name    = &kernel;
+            rtnl_reply.msg_namelen = sizeof(kernel);
+
+            /**
+             * Read as much data as fits in the receive buffer
+             */
+            if ((length = recvmsg(fd, &rtnl_reply, 0)) != 0) {
+                for (msg_ptr = (struct nlmsghdr *)reply; NLMSG_OK(msg_ptr, length);
+                     msg_ptr = NLMSG_NEXT(msg_ptr, length)) {
+                    switch (msg_ptr->nlmsg_type) {
+                    case NLMSG_DONE:
+                        /**
+                         * This is the special meaning NLMSG_DONE message we asked for by using NLM_F_DUMP flag
+                         */
+                        msg_done = 1;
+                        break;
+                    case RTM_NEWLINK:
+                        /**
+                         * This is a RTM_NEWLINK message, which contains lots of information about a link
+                         */
+                        if (get_link_metrics(msg_ptr, local_interface_name, link_metrics)) {
+                            msg_done = 1;
+                            result   = true;
+                        }
+                        break;
+                    }
+                }
+            }
+        }
+    }
+
+    return result;
+}
+
+static bool get_link_metrics(const std::string &local_interface_name, sLinkMetrics &link_metrics)
```

Comment:
I see now this is the 4th function overload with the same `get_link_metrics` name. This is fine, but can we at least add doxygen documentation to each function?

Reply:
Link metrics cannot be obtained with `base_wlan_hal_nl80211::send_nl80211_msg` as is. See https://github.com/prplfoundation/prplMesh/pull/831#discussion_r386899613

When you say BCL, I guess you mean BPL. Be aware that in #900 I've implemented the classes to send NL80211 commands and all Netlink related stuff in BWL. First of all we should decide  where those classes should be. 

I've documented the `get_link_metrics` functions.